File: rs/nonces/src/pool.rs
use std::collections::HashMap;
use std::time::{Instant, SystemTime, UNIX_EPOCH};

use alloy::primitives::{Address, FixedBytes, U256};
use alloy::providers::{Provider, ProviderBuilder};

use crate::constants::MAX_U256;
use crate::error::NonceError;
use crate::rpc::get_nonce_from_precompile;
use crate::types::{NonceMode, NoncePoolOptions, NoncePoolStats, NonceSlot, SlotState};

/// Production-grade nonce pool for Tempo's 2D nonce system.
///
/// Supports two modes:
/// - **Lanes**: N independent parallel nonce sequences (nonceKey 1..N)
/// - **Expiring**: Single TIP-1009 slot (nonceKey = MAX_U256) with time-bounded validity
///
/// # State Machine
///
/// ```text
/// free → acquire() → reserved → submit() → submitted → confirm() → free (nonce++)
///                                         → fail()    → free
///                   → fail()    → free
///                   → reap()    → free (after TTL)
/// any  → release() → free
/// ```
///
/// # Thread safety
///
/// `NoncePool` requires `&mut self` for all mutation methods and is therefore
/// `!Sync`. For shared access across async tasks, wrap in `Arc<Mutex<NoncePool>>`.
pub struct NoncePool {
    address: Address,
    rpc_url: String,
    mode: NonceMode,
    lane_count: u32,
    reservation_ttl_ms: u64,
    valid_before_offset_s: u64,
    chain_id: u64,
    validate_chain_id: bool,

    slots: Vec<NonceSlot>,
    slot_index: HashMap<U256, usize>,
    initialized: bool,

    confirmed_count: u64,
    failed_count: u64,
    reaped_count: u64,
}

impl std::fmt::Debug for NoncePool {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("NoncePool")
            .field("address", &self.address)
            .field("mode", &self.mode)
            .field("lane_count", &self.lane_count)
            .field("initialized", &self.initialized)
            .field("slot_count", &self.slots.len())
            .finish()
    }
}

impl NoncePool {
    /// Create a new nonce pool. Validates options but does NOT call RPC.
    /// Call [`init()`](Self::init) to fetch initial nonces.
    pub fn new(options: NoncePoolOptions) -> Result<Self, NonceError> {
        if options.address == Address::ZERO {
            return Err(NonceError::MissingAddress);
        }
        if options.rpc_url.is_empty() {
            return Err(NonceError::MissingRpcUrl);
        }
        if options.lanes < 1 {
            return Err(NonceError::InvalidLanes);
        }
        if options.reservation_ttl_ms == 0 {
            return Err(NonceError::InvalidTtl);
        }
        if options.valid_before_offset_s == 0 {
            return Err(NonceError::InvalidValidBefore);
        }

        Ok(Self {
            address: options.address,
            rpc_url: options.rpc_url,
            mode: options.mode,
            lane_count: options.lanes,
            reservation_ttl_ms: options.reservation_ttl_ms,
            valid_before_offset_s: options.valid_before_offset_s,
            chain_id: options.chain_id,
            validate_chain_id: options.validate_chain_id,
            slots: Vec::new(),
            slot_index: HashMap::new(),
            initialized: false,
            confirmed_count: 0,
            failed_count: 0,
            reaped_count: 0,
        })
    }

    /// Configured chain ID.
    pub fn chain_id(&self) -> u64 {
        self.chain_id
    }

    /// Initialize the pool by querying on-chain nonce values.
    pub async fn init(&mut self) -> Result<(), NonceError> {
        if self.initialized {
            return Err(NonceError::AlreadyInitialized);
        }
        self.fetch_and_init_slots().await?;
        self.initialized = true;
        Ok(())
    }

    async fn fetch_and_init_slots(&mut self) -> Result<(), NonceError> {
        let provider = ProviderBuilder::new().connect_http(
            self.rpc_url
                .parse()
                .map_err(|_| NonceError::MissingRpcUrl)?,
        );

        if self.validate_chain_id {
            let actual: u64 = provider
                .get_chain_id()
                .await
                .map_err(|e| NonceError::Rpc(alloy::contract::Error::TransportError(e)))?;
            if actual != self.chain_id {
                return Err(NonceError::ChainIdMismatch {
                    configured: self.chain_id,
                    actual,
                });
            }
        }

        self.slots.clear();
        self.slot_index.clear();

        match self.mode {
            NonceMode::Lanes => {
                let futs: Vec<_> = (1..=self.lane_count)
                    .map(|i| {
                        let key = U256::from(i);
                        let prov = &provider;
                        let addr = self.address;
                        async move {
                            let nonce = get_nonce_from_precompile(prov, addr, key).await?;
                            Ok::<_, NonceError>((key, nonce))
                        }
                    })
                    .collect();
                let results = futures::future::join_all(futs).await;
                for (idx, result) in results.into_iter().enumerate() {
                    let (key, nonce) = result?;
                    self.slots.push(NonceSlot::new(key, nonce));
                    self.slot_index.insert(key, idx);
                }
            }
            NonceMode::Expiring => {
                let key = MAX_U256;
                let nonce = get_nonce_from_precompile(&provider, self.address, key).await?;
                self.slots.push(NonceSlot::new(key, nonce));
                self.slot_index.insert(key, 0);
            }
        }

        Ok(())
    }

    /// Acquire a free slot, transitioning it to Reserved.
    ///
    /// If `request_id` is provided and a slot with the same ID exists in
    /// Reserved or Submitted state, that slot is returned (idempotency).
    pub fn acquire(&mut self, request_id: Option<&str>) -> Result<&NonceSlot, NonceError> {
        if !self.initialized {
            return Err(NonceError::NotInitialized);
        }

        if let Some(rid) = request_id {
            if let Some(idx) = self.find_slot_by_request_id(rid) {
                return Ok(&self.slots[idx]);
            }
        }

        self.reap();

        let idx = self
            .slots
            .iter()
            .position(|s| s.state == SlotState::Free)
            .ok_or(NonceError::Exhausted)?;

        let now = Instant::now();
        let slot = &mut self.slots[idx];
        slot.state = SlotState::Reserved;
        slot.reserved_at = Some(now);
        slot.request_id = request_id.map(String::from);

        if self.mode == NonceMode::Expiring {
            let unix_now = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs();
            slot.valid_before = Some(unix_now + self.valid_before_offset_s);
        }

        Ok(&self.slots[idx])
    }

    /// Transition a slot from Reserved to Submitted.
    pub fn submit(&mut self, nonce_key: U256, tx_hash: FixedBytes<32>) -> Result<(), NonceError> {
        let slot = self.get_slot_mut(nonce_key)?;

        if slot.state != SlotState::Reserved {
            return Err(NonceError::InvalidState {
                action: "submit",
                nonce_key,
                actual: slot.state.as_str(),
                expected: "reserved",
            });
        }

        slot.state = SlotState::Submitted;
        slot.submitted_at = Some(Instant::now());
        slot.tx_hash = Some(tx_hash);
        Ok(())
    }

    /// Confirm a submitted slot. Increments nonce by 1 and returns to Free.
    pub fn confirm(&mut self, nonce_key: U256) -> Result<(), NonceError> {
        let slot = self.get_slot_mut(nonce_key)?;

        if slot.state != SlotState::Submitted {
            return Err(NonceError::InvalidState {
                action: "confirm",
                nonce_key,
                actual: slot.state.as_str(),
                expected: "submitted",
            });
        }

        slot.nonce += 1;
        slot.reset();
        self.confirmed_count += 1;
        Ok(())
    }

    /// Fail a reserved or submitted slot. Nonce unchanged, returns to Free.
    pub fn fail(&mut self, nonce_key: U256) -> Result<(), NonceError> {
        let slot = self.get_slot_mut(nonce_key)?;

        if slot.state != SlotState::Reserved && slot.state != SlotState::Submitted {
            return Err(NonceError::InvalidState {
                action: "fail",
                nonce_key,
                actual: slot.state.as_str(),
                expected: "submitted\" or \"reserved",
            });
        }

        slot.reset();
        self.failed_count += 1;
        Ok(())
    }

    /// Release any slot back to Free, regardless of state.
    pub fn release(&mut self, nonce_key: U256) -> Result<(), NonceError> {
        let slot = self.get_slot_mut(nonce_key)?;
        slot.reset();
        Ok(())
    }

    /// Reap stale reserved slots that exceed the TTL.
    /// Returns snapshots of reaped slots (state before reset).
    pub fn reap(&mut self) -> Vec<NonceSlot> {
        let now = Instant::now();
        let ttl_ms = self.reservation_ttl_ms;
        let mut reaped = Vec::new();

        for slot in &mut self.slots {
            if slot.state == SlotState::Reserved {
                if let Some(reserved_at) = slot.reserved_at {
                    if now.duration_since(reserved_at).as_millis() as u64 > ttl_ms {
                        reaped.push(slot.clone());
                        slot.reset();
                    }
                }
            }
        }

        self.reaped_count += reaped.len() as u64;
        reaped
    }

    /// Read-only view of all slots.
    pub fn slots(&self) -> &[NonceSlot] {
        &self.slots
    }

    /// Current aggregate statistics.
    pub fn stats(&self) -> NoncePoolStats {
        let mut free = 0;
        let mut reserved = 0;
        let mut submitted = 0;

        for slot in &self.slots {
            match slot.state {
                SlotState::Free => free += 1,
                SlotState::Reserved => reserved += 1,
                SlotState::Submitted => submitted += 1,
            }
        }

        NoncePoolStats {
            total: self.slots.len(),
            free,
            reserved,
            submitted,
            confirmed: self.confirmed_count,
            failed: self.failed_count,
            expired: self.reaped_count,
        }
    }

    /// Re-query all on-chain nonces and reset slots to Free.
    /// Preserves cumulative stats.
    /// Nonce fetches are issued in parallel for all slots.
    pub async fn reset(&mut self) -> Result<(), NonceError> {
        if !self.initialized {
            return Err(NonceError::NotInitialized);
        }

        let provider = ProviderBuilder::new().connect_http(
            self.rpc_url
                .parse()
                .map_err(|_| NonceError::MissingRpcUrl)?,
        );

        let futs: Vec<_> = self
            .slots
            .iter()
            .map(|slot| get_nonce_from_precompile(&provider, self.address, slot.nonce_key))
            .collect();
        let results = futures::future::join_all(futs).await;
        for (slot, result) in self.slots.iter_mut().zip(results) {
            slot.nonce = result?;
            slot.reset();
        }

        Ok(())
    }

    fn get_slot_mut(&mut self, nonce_key: U256) -> Result<&mut NonceSlot, NonceError> {
        let idx = *self
            .slot_index
            .get(&nonce_key)
            .ok_or(NonceError::SlotNotFound(nonce_key))?;
        Ok(&mut self.slots[idx])
    }

    fn find_slot_by_request_id(&self, request_id: &str) -> Option<usize> {
        self.slots.iter().position(|s| {
            (s.state == SlotState::Reserved || s.state == SlotState::Submitted)
                && s.request_id.as_deref() == Some(request_id)
        })
    }

    /// Create an initialized pool for testing without RPC.
    #[doc(hidden)]
    pub fn new_for_testing(
        mode: NonceMode,
        lane_count: u32,
        reservation_ttl_ms: u64,
        valid_before_offset_s: u64,
    ) -> Self {
        let mut pool = Self {
            address: Address::repeat_byte(0x01),
            rpc_url: "http://localhost:8545".to_string(),
            mode,
            lane_count,
            reservation_ttl_ms,
            valid_before_offset_s,
            chain_id: crate::constants::MODERATO_CHAIN_ID,
            validate_chain_id: false,
            slots: Vec::new(),
            slot_index: HashMap::new(),
            initialized: true,
            confirmed_count: 0,
            failed_count: 0,
            reaped_count: 0,
        };

        match mode {
            NonceMode::Lanes => {
                for i in 1..=lane_count {
                    let key = U256::from(i);
                    let idx = pool.slots.len();
                    pool.slots.push(NonceSlot::new(key, 0));
                    pool.slot_index.insert(key, idx);
                }
            }
            NonceMode::Expiring => {
                let key = MAX_U256;
                pool.slots.push(NonceSlot::new(key, 0));
                pool.slot_index.insert(key, 0);
            }
        }

        pool
    }
}

File: ts/packages/nonces/src/pool.ts
import { createPublicClient, http, isAddress } from "viem";
import type { PublicClient } from "viem";
import type { NonceSlot, NoncePoolOptions, NoncePoolStats, NonceMode } from "./types";
import { getNonceFromPrecompile } from "./rpc";
import {
  MODERATO_CHAIN_ID,
  DEFAULT_LANES,
  DEFAULT_RESERVATION_TTL_MS,
  DEFAULT_VALID_BEFORE_OFFSET_S,
  MAX_UINT256,
} from "./constants";

/**
 * Nonce pool for Tempo's 2D nonce system.
 * Lifecycle: `free → reserved → submitted → confirmed → free`.
 */
export class NoncePool {
  private readonly address: `0x${string}`;
  private readonly rpcUrl: string;
  private readonly mode: NonceMode;
  private readonly laneCount: number;
  private readonly reservationTtlMs: number;
  private readonly validBeforeOffsetS: number;
  // Prefixed to avoid collision with the public chainId getter below.
  private readonly _chainId: number;
  private readonly validateChainId: boolean;

  private client: PublicClient | null = null;
  private slots: Map<bigint, NonceSlot> = new Map();
  private initialized = false;
  private initializing = false;
  private confirmedCount = 0;
  private failedCount = 0;
  private reapedCount = 0;

  constructor(options: NoncePoolOptions) {
    if (!isAddress(options.address)) throw new Error("NoncePool: invalid address format");
    if (!options.rpcUrl) throw new Error("NoncePool: rpcUrl is required");

    if (options.mode === "expiring" && options.lanes && options.lanes > 1) {
      throw new Error("NoncePool: lanes option is not supported in expiring mode");
    }

    this.address = options.address;
    this.rpcUrl = options.rpcUrl;
    this.mode = options.mode ?? "lanes";
    this.laneCount = options.lanes ?? DEFAULT_LANES;
    this.reservationTtlMs = options.reservationTtlMs ?? DEFAULT_RESERVATION_TTL_MS;
    this.validBeforeOffsetS = options.validBeforeOffsetS ?? DEFAULT_VALID_BEFORE_OFFSET_S;
    this._chainId = options.chainId ?? MODERATO_CHAIN_ID;
    this.validateChainId = options.validateChainId ?? false;

    if (!Number.isInteger(this.laneCount) || this.laneCount < 1)
      throw new Error("NoncePool: lanes must be an integer >= 1");
    if (this.reservationTtlMs <= 0) throw new Error("NoncePool: reservationTtlMs must be > 0");
    if (this.validBeforeOffsetS <= 0) throw new Error("NoncePool: validBeforeOffsetS must be > 0");
  }

  /** Chain ID this pool is configured for. */
  get chainId(): number {
    return this._chainId;
  }

  /**
   * Initialize the pool by querying on-chain nonce values.
   * Must be called before `acquire()`.
   *
   * **Note:** The `chainId` option is tracked locally and is not validated
   * against the RPC endpoint unless `validateChainId: true` is passed to the
   * constructor. Without validation, a mismatched `rpcUrl` silently returns
   * incorrect nonce values.
   *
   * @throws If already initialized (call `reset()` to re-sync)
   * @throws If initialization is already in progress (concurrent call)
   * @throws If `validateChainId` is true and the RPC chain ID does not match
   */
  async init(): Promise<void> {
    if (this.initialized) {
      throw new Error("NoncePool: already initialized — call reset() to re-sync nonces");
    }
    if (this.initializing) {
      throw new Error("NoncePool: initialization already in progress");
    }
    this.initializing = true;
    try {
      this.client = createPublicClient({ transport: http(this.rpcUrl) });
      if (this.validateChainId) {
        const rpcChainId = await this.client.getChainId();
        if (rpcChainId !== this._chainId) {
          throw new Error(
            `NoncePool: chainId mismatch — configured ${this._chainId}, RPC returned ${rpcChainId}`,
          );
        }
      }
      this.slots = this._slotsToMap(await this._fetchInitialSlots());
      this.initialized = true;
    } finally {
      this.initializing = false;
    }
  }

  /**
   * Reserve the next free slot. O(N) linear scan over slots.
   *
   * @param requestId - Optional idempotency key. If a reserved or submitted slot with this
   * requestId exists, returns it. After confirm/fail/reap, a new slot is allocated.
   * @returns The reserved NonceSlot (mutable reference — use `getSlots()` for snapshot copies)
   * @throws If no free slots are available (pool exhausted)
   * @throws If pool is not initialized
   */
  acquire(requestId?: string): Readonly<NonceSlot> {
    this._assertInitialized();

    if (requestId !== undefined) {
      for (const s of this.slots.values()) {
        if (s.requestId === requestId && (s.state === "reserved" || s.state === "submitted")) {
          return Object.freeze({ ...s });
        }
      }
    }

    this.reap();

    let free: NonceSlot | undefined;
    for (const s of this.slots.values()) {
      if (s.state === "free") {
        free = s;
        break;
      }
    }
    if (!free) {
      throw new Error("NoncePool: no free slots available");
    }

    const now = Date.now();
    free.state = "reserved";
    free.reservedAt = now;
    free.requestId = requestId;

    if (this.mode === "expiring") {
      free.validBefore = Math.floor(now / 1000) + this.validBeforeOffsetS;
    }

    return Object.freeze({ ...free });
  }

  /**
   * Mark a reserved slot as submitted (transaction sent to mempool).
   *
   * @param nonceKey - The lane key of the slot
   * @param txHash - Transaction hash
   * @throws If no slot exists for `nonceKey`
   * @throws If slot is not in "reserved" state
   */
  submit(nonceKey: bigint, txHash: `0x${string}`): void {
    const slot = this._getSlot(nonceKey);
    if (slot.state !== "reserved") {
      throw new Error(
        `Cannot submit slot ${nonceKey}: state is "${slot.state}", expected "reserved"`,
      );
    }
    slot.state = "submitted";
    slot.submittedAt = Date.now();
    slot.txHash = txHash;
  }

  /**
   * Mark a submitted slot as confirmed (transaction included in a block).
   *
   * Increments the nonce and resets the slot to "free", making it available
   * for the next transaction. In expiring mode, callers needing a fresh
   * on-chain nonce should call `reset()` before the next `acquire()`.
   *
   * @param nonceKey - The lane key of the slot
   * @throws If no slot exists for `nonceKey`
   * @throws If slot is not in "submitted" state
   */
  confirm(nonceKey: bigint): void {
    const slot = this._getSlot(nonceKey);
    if (slot.state !== "submitted") {
      throw new Error(
        `Cannot confirm slot ${nonceKey}: state is "${slot.state}", expected "submitted"`,
      );
    }
    if (this.mode === "lanes") {
      slot.nonce += 1n;
    }
    this.confirmedCount++;
    this._resetSlot(slot);
  }

  /**
   * Mark a slot as failed (transaction rejected by the network).
   *
   * Resets the slot to "free" with the same nonce value, allowing the caller
   * to retry with the same nonce (the nonce was not consumed on-chain).
   *
   * Also accepts `reserved` state (cancels a reservation that was never
   * submitted). Unlike `release()`, this increments `failedCount` in stats.
   *
   * @param nonceKey - The lane key of the slot
   * @throws If no slot exists for `nonceKey`
   * @throws If slot is not in "submitted" or "reserved" state
   */
  fail(nonceKey: bigint): void {
    const slot = this._getSlot(nonceKey);
    if (slot.state !== "submitted" && slot.state !== "reserved") {
      throw new Error(
        `Cannot fail slot ${nonceKey}: state is "${slot.state}", expected "submitted" or "reserved"`,
      );
    }
    this.failedCount++;
    this._resetSlot(slot);
  }

  /**
   * Release a slot back to "free" state regardless of current state.
   * Use for explicit cleanup or cancellation.
   *
   * @param nonceKey - The lane key of the slot
   * @throws If no slot exists for `nonceKey`
   */
  release(nonceKey: bigint): void {
    const slot = this._getSlot(nonceKey);
    this._resetSlot(slot);
  }

  /**
   * Reclaim slots that have exceeded the reservation TTL.
   *
   * @returns Snapshots of slots as they were just before being reaped
   */
  reap(): NonceSlot[] {
    this._assertInitialized();
    const now = Date.now();
    const reaped: NonceSlot[] = [];

    for (const slot of this.slots.values()) {
      if (
        slot.state === "reserved" &&
        slot.reservedAt !== undefined &&
        now - slot.reservedAt > this.reservationTtlMs
      ) {
        reaped.push({ ...slot });
        this._resetSlot(slot);
      }
    }

    this.reapedCount += reaped.length;
    return reaped;
  }

  /** Get a snapshot of all slots (shallow copies — safe to read, won't affect internal state). */
  getSlots(): readonly Readonly<NonceSlot>[] {
    return [...this.slots.values()].map((s) => ({ ...s }));
  }

  /** Get aggregate statistics. */
  getStats(): NoncePoolStats {
    let free = 0;
    let reserved = 0;
    let submitted = 0;
    for (const s of this.slots.values()) {
      if (s.state === "free") free++;
      else if (s.state === "reserved") reserved++;
      else if (s.state === "submitted") submitted++;
    }
    return {
      total: this.slots.size,
      free,
      reserved,
      submitted,
      confirmed: this.confirmedCount,
      failed: this.failedCount,
      expired: this.reapedCount,
    };
  }

  /**
   * Re-query all on-chain nonce values and reset all slots to "free".
   * Cumulative stats (confirmed/failed/expired) are preserved.
   * Call after a node restart or detected chain reorganization.
   *
   */
  async reset(): Promise<void> {
    this._assertInitialized();
    this.slots = this._slotsToMap(await this._fetchInitialSlots());
  }

  private _slotsToMap(slots: NonceSlot[]): Map<bigint, NonceSlot> {
    return new Map(slots.map((s) => [s.nonceKey, s]));
  }

  private async _fetchInitialSlots(): Promise<NonceSlot[]> {
    const slots: NonceSlot[] = [];

    if (this.mode === "lanes") {
      const promises: Promise<bigint>[] = [];
      for (let i = 1; i <= this.laneCount; i++) {
        promises.push(getNonceFromPrecompile(this.client!, this.address, BigInt(i)));
      }
      const nonces = await Promise.all(promises);
      for (let i = 0; i < this.laneCount; i++) {
        slots.push({
          nonceKey: BigInt(i + 1),
          nonce: nonces[i]!,
          state: "free",
        });
      }
    } else {
      const nonce = await getNonceFromPrecompile(this.client!, this.address, MAX_UINT256);
      slots.push({
        nonceKey: MAX_UINT256,
        nonce,
        state: "free",
      });
    }

    return slots;
  }

  private _getSlot(nonceKey: bigint): NonceSlot {
    const slot = this.slots.get(nonceKey);
    if (!slot) {
      throw new Error(`NoncePool: slot not found for nonceKey=${nonceKey}`);
    }
    return slot;
  }

  private _resetSlot(slot: NonceSlot): void {
    slot.state = "free";
    slot.reservedAt = undefined;
    slot.submittedAt = undefined;
    slot.txHash = undefined;
    slot.requestId = undefined;
    slot.validBefore = undefined;
  }

  private _assertInitialized(): void {
    if (!this.initialized) {
      throw new Error("NoncePool not initialized — call init() first");
    }
  }
}
